Assets/MixedRealityToolkit/Utilities/Editor/Setup/MixedRealityToolkitFiles.cs

```diff
@@ -246,5 +206,70 @@ public static string MapRelativeFilePath(MixedRealityToolkitModuleType module, s
         }
+
+        private static readonly Dictionary<string, MixedRealityToolkitModuleType> moduleNameMap = new Dictionary<string, MixedRealityToolkitModuleType>()
+        {
+            { "", MixedRealityToolkitModuleType.Core },
+            { "Generated", MixedRealityToolkitModuleType.Generated },
+            { "Providers", MixedRealityToolkitModuleType.Providers },
+            { "Services", MixedRealityToolkitModuleType.Services },
+            { "SDK", MixedRealityToolkitModuleType.SDK },
+            { "Examples", MixedRealityToolkitModuleType.Examples },
+            { "Tests", MixedRealityToolkitModuleType.Tests }
+        };
+
+        private static bool FindMatchingModule(string folder, out MixedRealityToolkitModuleType result)
+        {
+            // Matches an optional module suffix, e.g. ".Services"
+            const string modulePattern = @"(\.(?<module>\w+))?";
+            // Matches a version string, e.g. "2.0.0-20190611.2"
+            const string versionPattern = @"(?<version>[.\-0-9]+)";
+            // Matches the naming pattern in the MRTK repository
+            // e.g. "MixedRealityToolkit.Services"
+            const string mrtkPattern = @"MixedRealityToolkit" + modulePattern;
+            // Matches "Microsoft.MixedReality.Toolkit", followed by optional module name, followed by version number
+            // e.g.: "Microsoft.MixedReality.Toolkit.Services.2.0.0-20190611.2"
+            // This alternate path is used if above isn't found. This is to work around long paths issue with NuGetForUnity
+            // https://github.com/GlitchEnzo/NuGetForUnity/issues/246
+            const string nugetParentPattern = @"Microsoft\.MixedReality\.Toolkit" + modulePattern + @"\." + versionPattern;
+
+            var dirInfo = new DirectoryInfo(folder);
+
+            if (TryMatchFolderPattern(dirInfo.Name, mrtkPattern, out result))
+            {
+                return true;
+            }
+            else if (TryMatchFolderPattern(dirInfo.Parent.Name, nugetParentPattern, out result))
+            {
+                return true;
+            }
+            else
+            {
+                result = MixedRealityToolkitModuleType.Core;
+                return false;
+            }
+        }
+
+        private static bool TryMatchFolderPattern(string name, string pattern, out MixedRealityToolkitModuleType result)
+        {
+            var folderMatches = System.Text.RegularExpressions.Regex.Matches(name, pattern);
+            if (folderMatches.Count == 1)
+            {
+                var moduleName = folderMatches[0].Groups["module"].Value;
+                if (moduleNameMap.TryGetValue(moduleName, out result))
+                {
+                    Debug.Log($"Found module path for {result}: {name}");
+                    return true;
+                }
+                else
+                {
+                    Debug.LogWarning($"Found MRTK folder with unknown module extension \"{moduleName}\"");
```

Comment:
We should probably have some other build/CI time test that validates that our modules/folders adhere to this - this warning is helpful but it's a bit late (i.e. imagine we add a new folder without updating the Dictionary above, we wouldn't find out until way after the build has completed (in fact, things would probably fail in ways that aren't obvious until we open the console and see this warning, and then go back from that).

This is a larger ask than what can happen in this PR and we should absolutely keep this, just wanted to raise that there's a fragility here that can and should be addressed earlier in the cycle.

Reply:
Maybe my other comment about dictionary completeness will actually handle this, soooo maybe nevermind :)